config/watcher.go

```diff
@@ -132,26 +134,28 @@ func newFSWatcher(ctx context.Context, configPath string, logger golog.Logger) (
 				return
 			case event := <-fsWatcher.Events:
 				if event.Op&fsnotify.Write == fsnotify.Write {
-					//nolint:gosec
-					rd, err := os.ReadFile(configPath)
-					if err != nil {
-						logger.Errorw("error reading config file after write", "error", err)
-						continue
-					}
-					if bytes.Equal(rd, lastRd) {
-						continue
-					}
-					lastRd = rd
-					newConfig, err := FromReader(cancelCtx, configPath, bytes.NewReader(rd), logger)
-					if err != nil {
-						logger.Errorw("error reading config after write", "error", err)
-						continue
-					}
-					select {
-					case <-cancelCtx.Done():
-						return
-					case configCh <- newConfig:
-					}
+					debounced(func() {
+						//nolint:gosec
+						rd, err := os.ReadFile(configPath)
+						if err != nil {
+							logger.Errorw("error reading config file after write", "error", err)
+							return
+						}
+						if bytes.Equal(rd, lastRd) {
+							return
+						}
+						lastRd = rd
+						newConfig, err := FromReader(cancelCtx, configPath, bytes.NewReader(rd), logger)
+						if err != nil {
+							logger.Errorw("error reading config after write", "error", err)
+							return
+						}
+						select {
+						case <-cancelCtx.Done():
+							return
+						case configCh <- newConfig:
+						}
+					})
```

Comment:
This was a driveby for a bug. Previously the filewatcher would see my IDE save the file, but it'd trigger on the very first write operation, whereas it actually takes many writes to fully put the file to disk. When that happened, it'd read before the file was complete, and get EOF errors and broken configs. This makes sure it only tries to read after no writes are seen for 500ms.

Reply:
Ah, got it